misc-scripts/xref_mapping/XrefParser/RefSeqCoordinateParser.pm

```diff
@@ -22,406 +22,317 @@ package XrefParser::RefSeqCoordinateParser;
-          $score = $transcript_result{$tid};
-          $tl_score = $tl_transcript_result{$tid};
-          if ($score > $transcript_score_threshold || $tl_score > $tl_transcript_score_threshold) {
-            if ($tl_score >= $best_tl_score) {
-              if ($tl_score > $best_tl_score) {
-                $best_id = $tid;
-                $best_score = $score;
-                $best_tl_score = $tl_score;
-              } elsif ($tl_score == $best_tl_score) {
-                if ($score > $best_score) {
-                  $best_id = $tid;
-                  $best_score = $score;
-                }
-              }
-            } elsif ($score >= $best_score) {
-              $best_id = $tid;
-              $best_score = $score;
-            }
-          }
-        }
+        my ($best_id, $best_score, $best_tl_score) = $self->compute_best_scores($transcript_result, $tl_transcript_result);
 
-# If a best match was defined for the refseq transcript, store it as direct xref for ensembl transcript
+        # If a best match was defined for the refseq transcript, store it as direct xref for ensembl transcript
         if ($best_id) {
-          my ($acc, $version) = split(/\./, $id);
-	  $version =~ s/\D//g if $version;
-          my $source_id;
-          $source_id = $mrna_source_id if $acc =~ /^NM_/;
-          $source_id = $ncrna_source_id if $acc =~ /^NR_/;
-          $source_id = $pred_mrna_source_id if $acc =~ /^XM_/;
-          $source_id = $pred_ncrna_source_id if $acc =~ /^XR_/;
-          # Accession should be of format NM_/XM_/NR_/XR_ otherwise it is not valid
-          if (!defined $source_id) { next; }
-          my $xref_id = $self->add_xref({ acc => $acc,
-                                          version => $version,
-                                          label => $id,
-                                          desc => undef,
-                                          source_id => $source_id,
-                                          species_id => $species_id,
-                                          dbi => $dbi,
-                                          info_type => 'DIRECT' });
-          $self->add_direct_xref($xref_id, $best_id, "Transcript", "", $dbi);
-
-	  my $t_of = $transcript_of;
-          my $g_of = $t_of->get_Gene();
-          my $entrez_id = $g_of->stable_id();
-          my $tl_of = $t_of->translation();
+          my ($acc, $version) = split(/\./x, $id);
+
+          my $source = $self->source_id_from_acc($acc);
+
+          next TRANSCRIPT unless defined $source;
+
+          my $xref_id = $self->add_xref({
+            acc        => $acc,
+            version    => $version,
+            label      => $id,
+            source_id  => $source,
+            species_id => $species_id,
+            dbi        => $dbi,
+            info_type  => 'DIRECT'
+          });
+          $self->add_direct_xref($xref_id, $best_id, 'Transcript', undef, $dbi);
+
+          my $entrez_id = $gene_of->stable_id;
```

Comment:
We need the same check here as for the transcript->stable_id
In some cases, the stable_id will be a randomly generated ID and the NCBIGene accession is stored as the display_xref
For example in homo_sapiens_otherfeatures_95_38
select stable_id, x.dbprimary_acc from gene g, xref x where display_xref_id = xref_id and analysis_id = 8360
+-----------+---------------+
| stable_id | dbprimary_acc |
+-----------+---------------+
| gene58752 | 3139          |
| gene58319 | 5555          |
| gene58242 | 6046          |
+-----------+---------------+